examples/hosts/app-integration/schema-upgrade/src/dataTransform.ts

```diff
@@ -0,0 +1,104 @@
+}
+
+/**
+ * Parse string data in version:one format into an array of simple objects that are easily imported into an
+ * inventory list.
+ * @param stringData - version:one formatted string data
+ * @returns An array of objects, each representing a single inventory item
+ */
+export function parseStringDataVersionOne(stringData: string) {
+    const version = readVersion(stringData);
+    if (version !== "one") {
+        throw new Error(`Expected to parse version one, got version ${version}`);
+    }
+    const itemStrings = stringData.split("\n");
+    itemStrings.shift(); // remove version line
+    return itemStrings.map((itemString) => {
+        const [itemNameString, itemQuantityString] = itemString.split(":");
+        return { name: itemNameString, quantity: parseInt(itemQuantityString, 10) };
+    });
+}
+
+/**
+ * Parse string data in version:two format into an array of simple objects that are easily imported into an
+ * inventory list.
+ * @param stringData - version:two formatted string data
+ * @returns An array of objects, each representing a single inventory item
+ */
+export function parseStringDataVersionTwo(stringData: string) {
+    const version = readVersion(stringData);
+    if (version !== "two") {
+        throw new Error(`Expected to parse version two, got version ${version}`);
+    }
+    const itemStrings = stringData.split("\n");
+    itemStrings.shift(); // remove version line
+    return itemStrings.map((itemString) => {
+        const [itemNameString, itemQuantityString] = itemString.split("\t");
+        return { name: itemNameString, quantity: parseInt(itemQuantityString, 10) };
+    });
+}
+
+function parseStringData(stringData: string) {
+    const version = readVersion(stringData);
+    if (version === "one") {
+        return parseStringDataVersionOne(stringData);
+    } else if (version === "two") {
+        return parseStringDataVersionTwo(stringData);
+    } else {
+        throw new Error(`Don't know how to parse version ${version}`);
+    }
+}
+
+function transformToOne(stringData: string) {
+    const inventoryItems = parseStringData(stringData);
+    const inventoryItemStrings = inventoryItems.map((inventoryItem) => {
+        return `${ inventoryItem.name }:${ inventoryItem.quantity.toString() }`;
+    });
+    return `version:one\n${inventoryItemStrings.join("\n")}`;
+}
+
+function transformToTwo(stringData: string) {
+    const inventoryItems = parseStringData(stringData);
+    const inventoryItemStrings = inventoryItems.map((inventoryItem) => {
+        return `${ inventoryItem.name }\t${ inventoryItem.quantity.toString() }`;
+    });
+    return `version:two\n${inventoryItemStrings.join("\n")}`;
+}
+
+export const inventoryListDataTransformationCallback: DataTransformationCallback = async (
+    exportedData: unknown,
+    modelVersion: string,
+) => {
+    if (typeof exportedData !== "string") {
+        throw new Error("Unexpected data format");
+    }
+
+    if (modelVersion === "one") {
```

Comment:
Maybe a general comment - have you thought about generalizing transformation schema? Just for an example: Old FluidContainer, (with old schema) + some sort of declarative statement of transformation -> Upgraded container. Some extensibility points could be before/after transformation hooks on containers. Transformation syntax could allow expressing partial transformations etc. I did not think this through, its more of a general thought (and not necessarily relevant to this PR). Either way Im glad you are abstracting out transformation!

Reply:
I think that could potentially make sense for some cases, depending on the sorts of transformation the developer wanted.  It's easy to imagine how this would work for e.g. SharedMap -> SharedDirectory, but maybe harder to envision something like 3 SharedMaps -> 3 subdirectories for a single SharedDirectory.  I'll include this as a topic to consider as I look at the integration with FluidContainer.